Analog Comparator device driver API

bjarki-andreasen opened the pull request:
## Introduction
An analog comparator compares the voltages of two analog signals connected to its negative and positive inputs. If the voltage at the positive input is higher than the negative input, the comparator produces the output `true` (or `false` if the output is inverted).

This device driver API allows for reading the output from the comparator, and configuring a trigger on output change, which either invokes a callback or sets a flag to be checked and cleared by the application.

The devicetree is used for hardware configuration of the comparator and its surrounding circuit, this includes but is not limited to:
* positive input routing
* negative input routing
* power mode
* output routing
* hysteresis
* current (amps) sources on inputs
* combining two comparators to work as a window comparator

## Usecases covered by this API
Comparators are typically used with no interaction from the application. A comparator may be configured to disable a PWM output if the positive input exceeds the negative input (killing the output to a motor controller if current exceeds safe values for example). In these cases, the device driver configures the comparator on init and it is never interacted with again (outside of powering it down before shutdown).

Comparators may be used as wakeup sources, or voltage level crossed triggers, at which point the application will want to set a trigger and get a callback once triggered, in these cases, the comparator API is used by the application.

## Usecases not covered by this API
Users may want to reconfigure the comparator's hardware configuration at runtime, since this operation is highly vendor specific, vendors will have to provide their own device specific APIs which matches their hardware. Device driver specific APIs will be created with this PR to demonstrate how to runtime configure comparators as well.

## Note
This PR takes over from this previous comparator API proposal #63094

Hardware considered while designing this API
* STM32L5 comparators
* Microchip SAM-E70 comparator
* NRF COMP (COMParator)
* NRF LPCOMP Low Power COMParator)
* EFM32LG ACMP (Analog CoMParator)
* External comparator with its output connected to GPIO
* NXP ACMP (Analog CoMParator)

bjarki-andreasen-nordic committed:
drivers: Add comparator API

Add comparator API header.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: comparator: Add initial files

Add top level CMakeLists.txt entry and Kconfig options along with
userspace handlers for comparator API.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: comparator: Add nRF COMP device driver

Add nRF COMP device driver and remove deprecated bindings from
dts/bindings/sensor.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

henrikbrixandersen commented:
> ## Note
> This PR takes over from this previous comparator API proposal #63094

Please also take the following RFC as input: https://github.com/zephyrproject-rtos/zephyr/issues/59887
 
> Hardware considered while designing this API
> 
> * STM32L5 comparators
> * Microchip SAM-E70 comparator
> * NRF COMP (COMParator)
> * NRF LPCOMP Low Power COMParator)
> * EFM32LG ACMP (Analog CoMParator)
> * External comparator with its output connected to GPIO

I trust you have seen the drivers for various analog comparators already in the tree under the sensor API, e.g.:
- `drivers/sensor/nxp/mcux_lpcmp`
- `drivers/sensor/nxp/mcux_acmp`
- `drivers/sensor/ite/ite_vcmp_it8xxx2`

henrikbrixandersen commented:
Architecture WG 2024-07-23:
- @bjarki-andreasen presents the proposed analog comparator API and how it addresses the concerns regarding vendor-specific configuration interfaces raised against #63094.
- @bjarki-andreasen explains that the proposed API focuses on the functionality of analog comparator itself, not the functionality of all the surrounding components, as these are very vendor-specific. For applications where the configuration of the components surrounding the comparator is static, the configuration can be specified fully in driver specific devicetree bindings. The comparator API will then be fully generic between drivers for different comparators. Run-time configuration APIs will need to be driver-specific to account for the vendor-specific, surrounding components (multiplexers etc.).
- @henrikbrixandersen notes that it would be nice to see the existing comparator drivers currently under the sensors subsystem ported to this new API. This would provide examples of how the proposed driver-specific configuration APIs could look.

bjarki-andreasen-nordic committed:
drivers: comparator: add mcux acmp device driver

Add mcux SDK based kinetis acmp device driver implementing the
comparator device driver API.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
sensor: mcux_acmp: namespace driver and kconfigs

The mcux_acmp will get support by the comparator subsystem. To avoid
namespace clashes, namespace the driver, kconfigs and use the
MCUX_ACMP config solely to select the MCUX SDK driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: comparator: add shell

Add shell for comparator device drivers.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Rebased and added comparator shell :)

kartben commented on samples/drivers/comparator/callback/src/main.c:
```diff
@@ -0,0 +1,57 @@
+
+	ret = comparator_set_trigger(sample_comparator, COMPARATOR_TRIGGER_BOTH_EDGES);
+	if (ret < 0) {
+		printf("failed to set comparator trigger (ret = %i)\n", ret);
```
Since you've paid attention to this elsewhere, I'll suggest this nit: add a %s for "trigger"/"callback" part of the format string :)

kartben commented on samples/drivers/comparator/pending/README.rst:
```diff
@@ -0,0 +1,32 @@
+   :name: Comparator pending
+   :relevant-api: comparator_interface
+
+   Monitor the output of a comparator
```
Both samples have the same description?

bjarki-andreasen replied:
I think I may replace the samples with an "empty" sample so users can just use the shell to test with. WDYT?

henrikbrixandersen replied:
I'm typically not fond of "empty" samples just for showcasing a shell module.

How about adding shell documentation instead (see e.g. https://docs.zephyrproject.org/latest/hardware/peripherals/can/shell.html and https://docs.zephyrproject.org/latest/hardware/peripherals/eeprom/shell.html) with details on how to enable the shell module? Along with a proper test of the shell module (see e.g. https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/can/shell)?

bjarki-andreasen replied:
@henrikbrixandersen sounds good, I will add shell docs + tests and remove the sample :)

bjarki-andreasen-nordic committed:
drivers: comparator: add nRF LPCOMP device driver

Add nRF LPCOMP device driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Added support for nRF LPCOMP (low power comparator)

Next step is to replace the empty sample with shell docs and shell tests :)

bjarki-andreasen-nordic committed:
tests: drivers: build_all: add comparator test suite

Add comparator build_all test suite designed to test multiple
devicetree overlays and boards for each comparator device driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: drivers: build_all: comparator: add nrf_lpcomp

Add nrf_lpcomp to build_all comparator test suite.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: drivers: build_all: comparator: add mcux_acmp

Add mcux_acmp device driver to build_all test suite.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on drivers/comparator/comparator_nrf_lpcomp.c:
```diff
@@ -0,0 +1,441 @@
+		break;
+
+	case COMPARATOR_TRIGGER_RISING_EDGE:
+		shim_nrf_lpcomp_data0.event_mask = NRFX_LPCOMP_EVT_EN_UP_MASK;
```
The masks here should be updated to `NRF_LPCOMP_INT_*_MASK` to reflect the latest changes made to https://github.com/zephyrproject-rtos/hal_nordic/pull/205.

anangl commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,626 @@
+		break;
+
+	case COMPARATOR_TRIGGER_RISING_EDGE:
+		shim_nrf_comp_data0.event_mask = NRFX_COMP_EVT_EN_UP_MASK;
```
And here `NRF_COMP_INT_*_MASK` should be used.

bjarki-andreasen-nordic committed:
drivers: comparator: add fake comparator

Add fake comparator driver and bindings for use with testing.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
## Update

* Added build_all test suite for comparators, specifically built to be used for testing various combinations of initial devicetree configurations.
* Fixed some issues discovered by the build_all test suite
* Added fake driver for comparator to serve as the basis for the comparator shell test suite
* Updated the manifest and HAL PRs

The only missing step now is to finalize the shell test suite :)

decsny reviewed (dismissed):
bindings should not be defined based on different software interfaces, we should control which driver is used (sensor or comparator) with some other mechanism, such as kconfig, or whatever system the arch WG resolves on for instance based configuration in the future

decsny commented:
@fgrandel since you have an interest in this, this is another use case for the multi instance config scheme that I have seen many times in the past as well, ie where people make new bindings for new drivers for the same hardware (an exception where I think this is okay is when an existing binding is not appropriately defined for DT and is being deprecated)

bjarki-andreasen commented:
> bindings should not be defined based on different software interfaces, we should control which driver is used (sensor or comparator) with some other mechanism, such as kconfig, or whatever system the arch WG resolves on for instance based configuration in the future

In this case, `-sensor` is added to the binding for bacwards compatibility, to avoid break existing and downstream drivers which rely on the existing binding. The plan is to deprecate the sensor driver upstream, and with it the `-sensor` binding.

The new binding is written to be way closer to the hardware, and the existing sensor drivers could be refactored to use it instead, but since I at least would like to deprecate them, I see little benefit in adding breaking changes to the existing driver and old binding.

bjarki-andreasen-nordic committed:
dts: common: nordic: adjust comparator nodes

Adjust comparator nodes of nrf SoCs to exclude the unused
io-channel-cells property and simplify the comment describing how
to configure the comparator hardware block as COMP or LPCOMP for
SoCs which support this.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

fgrandel commented:
> @fgrandel this is another use case [...] where people make new bindings for new drivers for the same hardware

Thanks, @decsny , very useful example.

> when an existing binding is not appropriately defined for DT

I'd like to frame this a bit differently: This is another instance of the futile attempt to define _the_ appropriate driver-supporting hardware model forever (cf #77638): Currently we force every contributor to claim that her new model is more appropriate for DT (aka "much closer to hardware") and the previous was "partially inappropriate" until again being proven "partially inappropriate" by the next unexpected contribution. We should embrace the fact and deal with it in a structured way instead.

I'll comment further on this in #77638 to not spam this PR.

@bjarki-andreasen I hope you see that this implies no critique at all of your contribution, I really like it.

> The plan is to deprecate the sensor based driver upstream

We should **deprecate the sensor in the same change set** then, otherwise the introduction of an alternative model is confusing for downstream users. Plus we either need to **keep the old name stable or introduce a comment to the migration README** as existing DT trees/overlays will no longer work after this change.

@decsny Guess that if a deprecation is introduced then the separate compatible can be accepted until we have a better solution. That's what you've proposed yourself, right?

bjarki-andreasen-nordic committed:
tests: drivers: comparator: add shell test suite

Add test suite for comparator device driver shell.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on dts/bindings/comparator/nordic,nrf-lpcomp.yaml:
```diff
@@ -0,0 +1,86 @@
+    type: string
+    enum:
+      - "REF0"
+      - "REF1"
```
"REF0" and "REF1" are confusing, "AIN0" and "AIN1" should be used instead. And why not just add these two as values of `reference-selection`, instead having `reference-selection` and `external-reference` separately?

anangl commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+    type: string
+    enum:
+      - "SE"
+      - "DIFF"
```
Couldn't these values be "single-ended" and "differential"?

bjarki-andreasen replied:
yes, but then they would not be easily incorporated into the enums used in the device driver API, `COMP_NRF_COMP_<property>_<dt string>` -> `COMP_NRF_COMP_MAIN_MODE_SE`

anangl commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+include: base.yaml
+
+properties:
+  positive-input:
```
Shouldn't this property be required (and similarly `main-mode` and `speed-mode`, for example)?

bjarki-andreasen replied:
no, these are only required if the comparator driver is selected (`status = "okay";`) unlike interrupts and others which are required regardless. Otherwise we need to start adding often meaningless default values just to have something to put there :)

anangl replied:
I don't quite understand your reply. Required properties are only checked for nodes that have status "okay".

anangl commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+    enum:
+      - "NO_HYST"
+      - "40MV"
+      - "50MV"
```
These values look a bit weird. At first, I read them as 40 and 50 megavolts. ;-)

bjarki-andreasen replied:
is that not what it is? are we not making lightning? :D

bjarki-andreasen replied:
kept this, I really don't want to add lower case letters to these for consistency, maybe `50MILLIVOLT`?

anangl commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+      - "OFF"
+      - "IEN_2UA5"
+      - "IEN_5UA"
+      - "IEN_10UA"
```
Also these values are quite weird, especially "IEN_2UA5".

bjarki-andreasen replied:
yeah, but they do match the product spec, so I think it make sense to keep them since they are documented there, and in the device specific header in tree :)

bjarki-andreasen replied:
Changed the names to just `2UA5` ...

anangl commented on include/zephyr/drivers/comparator/nrf_comp.h:
```diff
@@ -0,0 +1,160 @@
+ * @retval 0 if successful
+ * @retval negative errno-code otherwise
+ */
+int comp_nrf_comp_configure_se(const struct device *dev,
+			       enum comp_nrf_comp_input input,
+			       enum comp_nrf_comp_speed_mode speed_mode,
+			       enum comp_nrf_comp_isource isource,
+			       enum comp_nrf_comp_ext_ref ext_ref,
+			       enum comp_nrf_comp_ref ref,
+			       uint8_t th_down,
+			       uint8_t th_up);
```
I think it would be better to use a structure instead of this bunch of parameters. This way, users would not be forced to provide a value for parameters they don't won't or cannot use (like `ext_ref` when an internal reference is used or `isource` on some SoCs). A structure can also be extended with a new field (if such need arises) without breaking the API.

bjarki-andreasen replied:
I will change it to be a structure :)

bjarki-andreasen replied:
Done

anangl commented on dts/common/nordic/nrf54h20.dtsi:
```diff
@@ -806,11 +806,10 @@
 
 			comp: comparator@983000 {
-				compatible = "nordic,nrf-comp";
+				compatible = "nordic,nrf54-comp";
```
Such compatible does not seem to be defined.

bjarki-andreasen replied:
correct, we don't have drivers for them yet

bjarki-andreasen replied:
Added support for them, turns out they can be covered by the same driver, `nrf-comp` :)

bjarki-andreasen commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+  positive-input:
+    type: string
+    enum:
+      - "INPUT_0"
```
Because this matches what the NRFX team has decided to use, which I'm trying to follow as closely as possible, on the presumption that there is likely a reason why they chose to use `INPUT_0` for example :)

bjarki-andreasen replied:
Done, now using AINx for both `psel` and `extrefsel`

bjarki-andreasen commented on include/zephyr/drivers/comparator.h:
```diff
@@ -0,0 +1,149 @@
+typedef int (*comparator_api_trigger_is_pending)(const struct device *dev);
+
+__subsystem struct comparator_driver_api {
+	comparator_api_get_output get_output;
+	comparator_api_set_trigger set_trigger;
+	comparator_api_set_trigger_callback set_trigger_callback;
+	comparator_api_trigger_is_pending trigger_is_pending;
```
yes

anangl replied:
As I already signaled in https://github.com/zephyrproject-rtos/zephyr/pull/63094#issuecomment-1754832391, I'm not convinced it is a good idea to force the use of PM if a user needs to start/stop a comparator manually. To me it's an abuse of PM.

bjarki-andreasen commented on dts/bindings/comparator/nordic,nrf-lpcomp.yaml:
```diff
@@ -0,0 +1,86 @@
+  reference-selection:
+    type: string
+    enum:
+      - "SUPPLY_1_8"
```
I will ask the NRFX team why they chose SUPPLY_ here, maybe we can easily change it to VDD_ which I will then do :)

bjarki-andreasen replied:
Updated to `VDD_`

bjarki-andreasen commented on west.yml:
```diff
@@ -188,7 +188,7 @@ manifest:
         - hal
     - name: hal_nordic
-      revision: 9717cc2845116ac1c4465b1eb8cc913513601e2a
+      revision: pull/205/head
```
will do :)

bjarki-andreasen commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+  external-reference:
+    type: string
+    enum:
+      - "EXT_REF_0"
```
I do agree, don't get me wrong :D AIN0 would be much nicer

bjarki-andreasen replied:
Done :) Now using `AINx`

decsny commented:
> > @fgrandel this is another use case [...] where people make new bindings for new drivers for the same hardware
> 
> Thanks, @decsny , very useful example.
> 
> > when an existing binding is not appropriately defined for DT
> 
> I'd like to frame this a bit differently: This is another instance of the futile attempt to define _the_ appropriate driver-supporting hardware model forever (cf #77638): Currently we force every contributor to claim that her new model is more appropriate for DT (aka "much closer to hardware") and the previous was "partially inappropriate" until again being proven "partially inappropriate" by the next unexpected contribution. We should embrace the fact and deal with it in a structured way instead.
> 
> I'll comment further on this in #77638 to not spam this PR.
> 
> @bjarki-andreasen I hope you see that this implies no critique at all of your contribution, I really like it.
> 
> > The plan is to deprecate the sensor based driver upstream
> 
> We should **deprecate the sensor in the same change set** then, otherwise the introduction of an alternative model is confusing for downstream users. Plus we either need to **keep the old name stable or introduce a comment to the migration README** as existing DT trees/overlays will no longer work after this change.
> 
> @decsny Guess that if a deprecation is introduced then the separate compatible can be accepted until we have a better solution. That's what you've proposed yourself, right?

issue here is exactly what you said - you would have to also deprecate the sensor driver if we are tying bindings to drivers like this, which in some cases makes sense, ie when you actually do intend to deprecate the old driver, and I understand this is bjarki's goal but right now I don't think people want their drivers deprecated, because not everybody has the time to write a new one for a new API, it's not something anybody maintaining these drivers committed to on their own

the example where it makes sense to deprecate an old binding IMO is only when 
a) the DT is egregiously misused BUT it's determined that a breaking change to it would be inappropriate due to the binding's popular, stable use among users, therefore a transition period is deemed required, or
b) there is actually a concrete commitment by someone to write a new driver with a new binding

in this case, (a) doesn't seem to be met because there is not a misuse of DT in most of these bindings as far as I can tell, and (b) I'm not confident is met either, who is going to rewrite all these drivers? 

I will say, I don't speak for NXP but I am in the upstream engagement team at NXP, and I tend to expect that if NXP drivers are deprecated and no commitment from anyone to write new ones, then this issue will probably be escalated to TSC by NXP if we don't have the resources to work on this and avoid deprecated warning before next release. That's not to say that I don't agree with the idea here to have a comparator API, but unless TSC/project decides on a mass collaborative change like hwmv2 then randomly deprecating other people's stuff doesn't seem appropriate to me

bjarki-andreasen commented:
@decsny

I'm certainly not aiming at "randomly deprecating other peoples stuff" with this PR (or any work I have done I would hope).

This PR is updating one particular binding for one particular comparator, for which there is a new device driver to go with it. To preserve the existing driver for that same comparator, the old binding is renamed, and kept.

Just like when the RTC API was introduced, all counter API based drivers, which still exist in tree, where not deprecated. We may at some point choose to deprecate them, but they live on, so to will the sensor API based driver.

The zephyr project is a collaborative, ever growing project. Everything in this project is subject to change; drivers, bindings, subsystems, socs, everything.

That being said, introducing a new device driver API, does not mean that every single device driver in-tree which does not use it has to be adapted to it, or face deprecation/deletion. Just like when the RTC API was introduced, comparator API based drivers will slowly be added (hopefully), alongside potentially existing in-tree drivers based on other APIs.

I really feel a we are making a mountain out of a mole hill here...

decsny commented:
> @decsny
> 
> I'm certainly not aiming at "randomly deprecating other peoples stuff" with this PR (or any work I have done I would hope).


what I was responding to was fgrandel's suggestion

> I really feel a we are making a mountain out of a molehill here...

I'll unblock the PR because maybe I am being pedantic here  and you're right there are replacement drivers in the PR, but I really want to avoid zephyr's use of DT getting any more out of control than it already is

bjarki-andreasen commented:
@decsny I did not consider deprecating properties in bindings, that is probably way better, let me see what I can do to keep just one binding without breaking to much (if anything) :)

bjarki-andreasen commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,131 @@
+      - "INPUT_6"
+      - "INPUT_7"
+      - "INPUT_VDD_DIV2"
+      - "INPUT_VDD_DIV5"
```
fixed, now `VDD_DIV2` and `VDDH_DIV5`

bjarki-andreasen commented:
Latest force push addresses most of @anangl comments regarding nrf drivers :)

henrikbrixandersen commented:
> Latest force push addresses most of @anangl comments regarding nrf drivers :) Now working on acmp

Please let me know if I can help with the NXP ACMP, development-wise or test-wise. We use it downstream.

bjarki-andreasen commented:
> > Latest force push addresses most of @anangl comments regarding nrf drivers :) Now working on acmp
> 
> Please let me know if I can help with the NXP ACMP, development-wise or test-wise. We use it downstream.

Absolutely, I will ping you once I and hopefully @decsny are happy with it the driver and bindings :)

bjarki-andreasen-nordic committed:
drivers: sensor: mcux_acmp: update dts binding and adapt driver

Update the devicetree binding for the nxp,kinetis-acmp comparator and
move the binding to dts/bindings/comparator.

The update to the binding includes:
- Remove unused io-channel-cells property
- Remove unused sensor-device include
- Adding missing properties dac config, discrete mode config, and
  input configs.
- Rename properties to exclude redundant vendor prefix since props in
  this binding are not inhereted, and as such, don't need to be
  namespaced.
- Deprecate the old names of the renamed properties

The sensor based device driver has been updated to support both the
deprecated and new property names. This allows it to use both
nxp,enable-sample and filter-enable-sample for example.

Additionally, remove the unused io-channel-cells properties from
in-tree nodes of compatible = "nxp,kinetis-acmp"

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
@decsny We now only have 1 binding for the `nxp,kinetis-acmp`. The properties which have been renamed have had their old name deprecated, and the sensor API based `mcux_acmp.c` device driver has been updated to be able to use either the new or deprecated property names interchangeably.

The reason I want to update their names is to exclude the vendor prefix from the properties, which are redundant when the props are not inherited, in which case namespacing them would make sense :)

bjarki-andreasen-nordic committed:
docs: hardware: peripherals: add comparator docs

Add documentation for the comparator peripheral and shell.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
## Update

* Added description to `nxp,kinetis-acmp` devicetree binding with examples
* Added documentation for comparator peripheral and shell
* Added help text to comparator shell commands
* Added nRF COMP and LPCOMP support for nRF54L and nRF54H
* Aligned names between nRF COMP and LPCOMP devicetree properties
* Rebased PR

bjarki-andreasen-nordic committed:
doc: release notes: added comparator release notes

Added comparator support to release notes for release 4.0

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
MAINTAINERS: add entry for comparator drivers

Add entry for comparator drivers

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
doc: releases: migration-guide: deprecated nxp,kinetis-acmp props

Add entyr mentioning the deprecation of the nxp, prefixed properties
in the nxp,kinetis-acmp dts binding.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
@henrikbrixandersen The MCUX ACMP device driver is now ready for testing :)

anangl commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,783 @@
+int comp_nrf_comp_configure_se(const struct device *dev,
+			       const struct comp_nrf_comp_se_config *config)
+{
+	nrfx_comp_config_t nrf;
```
It would be better to use an initializer here (and similarly in `comp_nrf_comp_configure_diff()` and `shim_nrf_comp_init()`) to make sure no field is left with some random value from stack.

bjarki-andreasen replied:
Added initalizer

anangl commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,783 @@
+	}
+#endif
+
+	return 0;
```
This callback should not return 0 for actions that it does not implement.

bjarki-andreasen replied:
This matches the current implementation of pm_device_runtime. I agree that pm_device_runtime should support -ENOTUP for these calls, but it expects 0 whether implemented or not :)

anangl replied:
After 740eba1341ab2456c97390ea71d4be5d9ff8bab7 got merged, it is no longer needed to pretend that `PM_DEVICE_ACTION_TURN_ON` is implemented.

bjarki-andreasen commented on dts/bindings/comparator/nordic,nrf-comp.yaml:
```diff
@@ -0,0 +1,132 @@
+            status = "okay";
+            psel = "AIN0";
+            refsel = "INT_1V2";
+            main-mode = "SE";
```
Moved main-mode to top of examples and properties dict

bjarki-andreasen commented:
## Update

* Addressed comments regarding COMP and LPCOMP drivers/bindings

@anangl Regarding power management. The following is a diagram of a comparator:

![image](https://github.com/user-attachments/assets/0da97945-456b-43c5-8b0d-a78d111a79a2)

You power it on, and it will actively compare the two inputs, and produce an output. This is power management. If the application needs it, it will resume it. If the application does not need it, and has enabled power management, the application can put it. "enable" is synonymous with resume, and "disable" is synonymous with suspend.

decsny reviewed (dismissed):
I don't understand the objection to @anangl about the start/stop API. All of these drivers as far as I can tell are written so that if PM_DEVICE is `n`, the comparator will be enabled unconditionally. Which is frankly ridiculous because on many SOC I know of, the comparator is the single thing that consumes the most power of anything on the chip, and leaving it on all the time would be very wasteful. Yes, you are right that PM could be used to control it nicely, I see how you can map active/suspended to enabled/disabled, but why is it required to build entire PM and device PM subsystems to use the comparator api in a reasonable way? I have not seen a convincing justification for that. Are we expecting every case where someone uses a comparator to also want to use PM subsystem ? If so, then why not make a dependency on it instead of having the alternative "always on" code when PM_DEVICE is disabled ?

bjarki-andreasen commented:
> I don't understand the objection to @anangl about the start/stop API. All of these drivers as far as I can tell are written so that if PM_DEVICE is `n`, the comparator will be enabled unconditionally. Which is frankly ridiculous because on many SOC I know of, the comparator is the single thing that consumes the most power of anything on the chip, and leaving it on all the time would be very wasteful. Yes, you are right that PM could be used to control it nicely, I see how you can map active/suspended to enabled/disabled, but why is it required to build entire PM and device PM subsystems to use the comparator api in a reasonable way? I have not seen a convincing justification for that. Are we expecting every case where someone uses a comparator to also want to use PM subsystem ? If so, then why not make a dependency on it instead of having the alternative "always on" code when PM_DEVICE is disabled ?

I will counter with the inverse argument: "Why would all device driver APIs have to include a bespoke, yet duplicate in functionality, start/stop API, regardless of whether an application cares about power management or not?"

It is a design flaw to add a duplicate API in all device drivers. Especially when power management requires some higher level subsystem to be usable; reference counting, dependency tracking etc.

A small note, only `PM_DEVICE` and I would argue `PM_DEVICE_RUNTIME` is required to enable power management, which would work with all devices in the system, not just comparators, busses, power domains, everything. It adds 4K ROM and 1K RAM on an n nrf5340dk. Its not that bad, and it scales well to all devices.

bjarki-andreasen commented:
> > I will counter with the inverse argument: "Why would all device driver APIs have to include a bespoke, yet duplicate in functionality, start/stop API, regardless of whether an application cares about power management or not?"
> 
> What if the API was only available for the case where the device PM is not enabled ?

That would be a contradiction IMO, not enabling power management is a way to simplify the driver and save RAM/ROM, why would we not give the user the option of excluding power management entirely? It is a valid usecase for wall powered stuff to just not bother with PM at all.

> > It is a design flaw to add a duplicate API in all device drivers. Especially when power management requires some higher level subsystem to be usable; reference counting, dependency tracking etc.

> I'm not sure I understand what is your point here, what I am saying is exactly this, why should a simple comparator driver rely on having a more complicated higher level subsystem enabled to use in a reasonable way?

It does not, the comparator can be used with only `PM_DEVICE` enabled, which simply exposes the power management API, the comparator driver implements the exact same functionality whether it uses the comparator API's enable/disable call or the PM_DEVICE callback RESUME/SUSPEND action. The point is that the PM_DEVICE RESUME/SUSPEND **can** be used with a higher level subsystem, contrary to the ad-hoc enable/disable/start/stop APIs.

> > A small note, only `PM_DEVICE`, and I would argue `PM_DEVICE_RUNTIME`, is required to enable power management, which would work with all devices in the system, not just comparators; busses, power domains, clocks, everything. It adds 4K ROM and 1K RAM (system workqueue stack) on an nrf5340dk. Its not that bad, and it scales well to all devices. Enabling only `PM_DEVICE`, which is the "real" cost here, adds 500B ROM, 32 bytes RAM
> 
> 1 KB RAM is a LOT on some platform that are specifically designed under certain constraints. For example, the NXP [MCXC141 ](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_mcxc141.dtsi) has an ACMP (to be enabled in zephyr still) only has 8 KB RAM on the chip. I doubt that all the users who want to use this comparator API on that platform to use ACMP would want to also enable >1KB of power management subsystem code, especially when the comparator driver itself is already >1KB. Even some of the platforms you changed the DT of in this PR like the ke1xz, and the ke1xf, have only 24 KB and 32KB of RAM, respectively.

I did show the "worst case scenario" where the runtime subsystem was enabled, just exposing the resume/suspend action only added a 32 bytes of RAM... Not entirely sure that is only due to the comparator driver, other things in the system may also implement PM_DEVICE making the number a bit inaccurate.

fabiobaltieri commented on include/zephyr/drivers/comparator.h:
```diff
@@ -0,0 +1,150 @@
+
+/** @cond INTERNAL_HIDDEN */
+
+typedef int (*comparator_api_get_output)(const struct device *dev);
```
should this be just "get"? like `gpio_pin_get`

bjarki-andreasen replied:
The reason I'm specifying output here is to not confuse it with the state of the inputs :) The output could be inverted for example (high if positive input is lower than negative input), so `_output` here hopefully prevents confusing the logical state with the output state :)

fabiobaltieri commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,771 @@
+#endif
+
+	default:
+		ret = -EINVAL;
```
how about `return -EINVAL` here, `return 0` at the end and drop the ret variable? one less thing to track, I think it makes these a bit easier to follow, same below, you actually do it already in some of the functions anyway

fabiobaltieri commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,771 @@
+	shim_nrf_comp_data0.callback = callback;
+	shim_nrf_comp_data0.user_data = user_data;
+
+	if (callback != NULL && atomic_test_and_clear_bit(&shim_nrf_comp_data0.triggered, 0)) {
+		callback(dev, user_data);
+	}
```
mmh the extra logic beyond `data->callback = cb` here seems like asking for inconsistent implementation between drivers, how about moving this to the api level in the `comparator.h` implementation?

bjarki-andreasen replied:
The issue here is that the underlying nrfx driver obfuscates the interrupt line, converting it to an event handler which this driver has to track. The expected behavior is that enabling the callback enables the interrupt, which, if pending/triggered, immediately triggers the irq, calling the callback. This is the case for the mcux_acmp driver/hardware, and indeed also for the nrf hardware. I would like to not have this rather dirty workaround for the nrf driver be part of the API :)

fabiobaltieri commented on drivers/comparator/comparator_nrf_lpcomp.c:
```diff
@@ -0,0 +1,490 @@
+		break;
+
+	default:
+		ret = -EINVAL;
```
same comment as in the other driver on the return style

fabiobaltieri commented on drivers/comparator/comparator_shell.c:
```diff
@@ -0,0 +1,279 @@
+
+static K_SEM_DEFINE(triggered_sem, 0, 1);
+
+static int get_device_from_str(const struct shell *sh,
```
temping to suggest that this and the timeout one should go in `subsys/shell/shell_utils.c`, just an idea, maybe a followup

bjarki-andreasen replied:
agreed, but as a followup :) I'm sure these additions can be added tons of places in the tree, we should do that in one PR :)

fabiobaltieri replied:
yup, cool

anangl commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,771 @@
+
+static int shim_nrf_comp_reconfigure(const nrfx_comp_config_t *nrf)
+{
+	shim_nrf_comp_stop();
+
+	(void)nrfx_comp_reconfigure(nrf);
+
+	if (shim_nrf_comp_is_resumed()) {
+		shim_nrf_comp_start();
+	}
```
If I understand it correctly, reconfiguration is allowed while the comparator is suspended. The problem here is that an assertion in `nrfx_comp_stop()` will fail if the comparator is already stopped. I guess stopping should also be conditional.

bjarki-andreasen replied:
it is :) `shim_nrf_comp_stop();` and `shim_nrf_comp_start();` check if the state is already met by tracking the `.started` member:
```
static void shim_nrf_comp_start(void)
{
	if (shim_nrf_comp_data0.started) {
		return;
	}

	nrfx_comp_start(shim_nrf_comp_data0.event_mask, 0);
	shim_nrf_comp_data0.started = true;
}
```

anangl replied:
Right. I somehow missed that additional logic in `shim_nrf_comp_stop()`.

bjarki-andreasen commented:
@fabiobaltieri @anangl Thank you for the reviews! will address shortly :)

bjarki-andreasen commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,771 @@
+BUILD_ASSERT(SHIM_NRF_COMP_DT_INST_PSEL(0) != COMP_NRF_COMP_PSEL_VDD_DIV2);
+#endif
+
+#ifndef COMP_PSEL_PSEL_VddDiv2
```
fixed

bjarki-andreasen commented on drivers/comparator/comparator_nrf_comp.c:
```diff
@@ -0,0 +1,771 @@
+#else
+	(void)shim_nrf_comp_diff_config_to_nrf(&shim_nrf_comp_config0, &nrf);
+#endif
+	(void)nrfx_comp_init(&nrf, shim_nrf_comp_event_handler);
```
Updated to check the return value, looking through the `nrfx_comp_init` function, it can only fail on invalid state or failing to acquire something called `prs` which to my knowledge is not enabled for zephyr. I can check the value just to be safe, but really it should always return SUCCESS

bjarki-andreasen replied:
updated to check the value

bjarki-andreasen commented on include/zephyr/drivers/comparator/nrf_comp.h:
```diff
@@ -0,0 +1,164 @@
+	enum comp_nrf_comp_sp_mode sp_mode;
+	/** Current source configuration */
+	enum comp_nrf_comp_isource isource;
+	/** Negative external input selection */
+	enum comp_nrf_comp_extrefsel extrefsel;
```
agreed, changed to "External reference selection"

bjarki-andreasen commented on include/zephyr/drivers/comparator/nrf_comp.h:
```diff
@@ -0,0 +1,164 @@
+};
+
+/**
+ * @brief Configure comparator in single-ended mode using internal reference
```
removed "using internal reference" part

pkunieck commented:
There is a problem running this test on some intel platforms, I created an issue #79918

kartben commented:
@bjarki-andreasen would be great to have a couple bullet points about the new driver class and drivers in the release notes :)

bjarki-andreasen commented:
@kartben See #81402
